docs: Add AWS IoT Core integration docs (#1823)
mendersoftware/mender-docs: 08.Server-integration/06.AWS-IoT-Core/docs.md

---
title: AWS IoT Core integration
taxonomy:
    category: docs
---

> Mender supports preparing devices to use AWS IoT Core. After installing Mender on the device it is automatically created in AWS IoT Core and applications running on the device get access to it. With Mender in place, you can simply start writing your AWS IoT application by reading the Thing's private key and certificate from a well-defined place on the device. Secondly, the Device Shadow from the AWS IoT Core becomes available to manage from Mender.

Reviewer: Found the paragraph a little difficult to read. I tried rephrasing a few things in the following suggestion:
```suggestion
Mender supports preparing devices to use AWS IoT Core. After installing Mender on the device it is automatically created in AWS IoT Core and applications running on the device get access to it. With Mender, you can start writing your AWS IoT application by letting Mender manage the provisioning of the device certificate keys for IoT Core. Moreover, the Device Shadow is integrated into the Mender ecosystem letting you control the desired state and read the reported of your devices in AWS. 
```

This integration is available in all Mender plans, as well as Mender Open Source.

!!!!! You need [Mender Configure](../../09.Add-ons/10.Configure/docs.md) to distribute the AWS IoT Core keys and certificates to the devices.

> !!! It is currently possible to have only one AWS IoT Core integration configured per Mender Organization (aka. tenant). More advanced integration scenarios can be configured via webhooks.

Reviewer: ```suggestion
!!! It is currently possible to have only one AWS IoT Core integration configured per Mender Organization. More advanced integration scenarios can be configured via webhooks.
```

> !!! After you enable the integration, devices in the AWS IoT Core will be created automatically after they are accepted in the Mender UI. Existing devices (in both systems) are left unchanged.

Reviewer: ```
!!! After enabling the integration ...
```

                "iot:CreateCertificateFromCsr",
                "iot:CreatePolicy",
                "iot:CreateThing",
                "iot:DeleteCertificate",
                "iot:DeletePolicy",
                "iot:DeleteThing",
                "iot:DescribeAccountAuditConfiguration",
                "iot:DescribeCertificate",
                "iot:DescribeEndpoint",
                "iot:DescribeThing",
                "iot:DetachThingPrincipal",
                "iot:GetIndexingConfiguration",
                "iot:ListBillingGroups",
                "iot:ListScheduledAudits",
                "iot:ListThingGroups",
                "iot:ListThingGroupsForThing",
                "iot:ListThingPrincipals",
                "iot:ListThingTypes",
                "iot:ListThings",
                "iot:UpdateCertificate"
            ],
            "Resource": "*"
        }
    ]
}
```

!!! `AWSIoTFullAccess` AWS managed policy has extended privileges and can be used for getting started and demo purposes, but not recommended to production usage.

TBD: pic

AWS IoT Core Policy needs to be created. The policy will be attached to all accepted devices in the Mender UI.

> TBD: pic

Reviewer: I would also like to see an example device policy, for example:
```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "iot:Connect"
            ],
            "Resource": [
                "arn:aws:iot:us-east-1:123456789012:client/${iot:Connection.Thing.ThingName}"
            ]
        },
        {
            "Effect": "Allow",
            "Action": [
                "iot:Publish"
            ],
            "Resource": [
                "arn:aws:iot:us-east-1:123456789012:topic/${iot:Connection.Thing.ThingName}/*"
            ]
        }
    ]
}
```

Author: It supposed to be on the picture, but we can also add a json.

## Configuring Mender to integrate with AWS IoT Core

> To connect your devices to AWS IoT Core, you need to setup the integration by providing an AWS IAM access key pair and specify AWS IoT Core Policy name. 

Reviewer 2: To connect your devices to AWS IoT Core, you need to set up the integration by providing an AWS IAM access key pair and specifying the AWS IoT Core Device Policy name.

Open the Mender UI and navigate to `Settings` -> `Integrations`:

TBD: pic

> Choose the AWS IoT Core integration and enter your IAM Key ID, Key Secret, Region and Device Policy Name:

Reviewer 2: ```suggestion
Choose the AWS IoT Core integration and enter your IAM Access Key ID, Secret Access Key, Region and Device Policy Name:
```

TBD: pic

> You can create an access key under `Security credentials` tab in AWS IAM Users menu:

Reviewer 2: ```suggestion
You can create the AWS access keys from the `Security credentials` tab in the AWS IAM Users menu:
```

TBD: pic

To attach IAM Policy to the AWS user in AWS Management Console go to `IAM` -> `Users` -> `<username>`, press `Add permissions` in `Permissions` tab and attach `AWSIoTFullAccess` existing policy directly.

> !!! The `AWSIoTFullAccess` policy has the required permissions, but you can also use a more limited account if desired.

Reviewer 2: ```suggestion
!!! The `AWSIoTFullAccess` policy has all the required permissions, but we encourage you to use a more limited IAM policy.
```

TBD: pic

## Connect your devices to AWS IoT Core automatically using Mender

Go to the `Devices` tab in Mender and accept a device. After it gets authorized in Mender, it will also show up in your AWS IoT Core:

!!! Both Mender and AWS IoT Core use the same Device ID. This makes it easy to cross-reference your devices between Mender and AWS IoT Core.

_The pictures from azure integration will be replaced_
![Accept device](../05.Azure-IoT-Hub/image_4.png)

TBD: pic

From now on, your device can send data to AWS IoT Core!

## Device lifecycle overview

> In sum, the lifecycle of an AWS IoT Core integrated device is:

Reviewer 2: ```suggestion
The lifecycle of an AWS IoT Core integrated device is:
```

1. At first, the device gets accepted - you can use various flows to achieve this state. See [device authentication](../../02.Overview/13.Device-authentication/docs.md) for more information.

> 2. Mender provisions the device in AWS IoT Core - as a result of this operation, the Mender server generates [AWS IoT Core X.509 client certificate](https://docs.aws.amazon.com/iot/latest/developerguide/x509-client-certs.html) and registers it in AWS IoT Core.

Reviewer 2: ```suggestion
2. Mender provisions the device in AWS IoT Core - as a result of this operation, the Mender server generates an [AWS IoT Core X.509 client certificate](https://docs.aws.amazon.com/iot/latest/developerguide/x509-client-certs.html) and registers it in AWS IoT Core.
```

> 3. The Mender [Configure add-on](../../09.Add-ons/10.Configure/docs.md) sets Thing's private key and certificate in file (`/var/lib/mender-configure/device-config.json`) on the device. We recommend creating a [Mender configuration script](../../09.Add-ons/10.Configure/01.Device-integration/docs.md) to reconfigure the AWS IoT application and restart it with its new certificate.

Reviewer 2: ```suggestion
3. The Mender [Configure add-on](../../09.Add-ons/10.Configure/docs.md) sets the Thing's private key and certificate in the file (`/var/lib/mender-configure/device-config.json`) on the device. We recommend creating a [Mender configuration script](../../09.Add-ons/10.Configure/01.Device-integration/docs.md) to reconfigure the AWS IoT application and restart it with these credentials.
```

4. From now on, device applications can read the private key and certificate from the file and communicate with AWS IoT Core.

> 5. If the device gets `rejected` or `dismissed` in Mender, it is automatically inactivates associated with the related Thing Certificate in AWS IoT Core.

Reviewer 2: ```suggestion
5. If the device gets `rejected` or `dismissed`, Mender will automatically inactivate the certificate associated with the Thing in AWS IoT Core.
```

> 6. If the device gets `decommissioned`, it is automatically `Deleted` from AWS IoT Core.

Reviewer 2: ```suggestion
6. If the device gets `decommissioned`, the Thing and the related resources, including the Certificate, are automatically `Deleted` from AWS IoT Core.
```

![Device lifecycle](../05.Azure-IoT-Hub/device_lifecycle.png)

## Device Shadow management

> As an additional integration point, the Mender server obtains the Device Shadow of the devices from the AWS IoT Core.

Reviewer 2: ```suggestion
As an additional integration point, the Mender server obtains the Device Shadow of the devices from AWS IoT Core.
```

This means you can see and manage the Device Shadow directly from Mender, together with everything else about the device.

![Device Twin flow](../05.Azure-IoT-Hub/azure-iot-device-twin-flow.png)

The Device Shadow is available through the Mender APIs and UI.

![Device Twin edit](../05.Azure-IoT-Hub/azure-iot-device-twin-edit.png)

## Role Based Access Control

!!!!! Role Based Access Control is only available in the Mender Enterprise plan.
!!!!! See [the Mender features page](https://mender.io/product/features?target=_blank)
!!!!! for an overview of all Mender plans and features.

*Admin* permission is required to set up the integration,

> i.e. set the connection string of the AWS IoT Core in Mender's Settings.

Reviewer 2: ```suggestion
i.e. set the AWS access keys, region and device policy name for AWS IoT Core in the Mender's Settings.
```

## Device lifecycle overview

The lifecycle of an AWS IoT Core integrated device is:

1. At first, the device gets accepted - you can use various flows to achieve this state. See [device authentication](../../02.Overview/13.Device-authentication/docs.md) for more information.
2. Mender provisions the device in AWS IoT Core - as a result of this operation, the Mender server generates an [AWS IoT Core X.509 client certificate](https://docs.aws.amazon.com/iot/latest/developerguide/x509-client-certs.html) and registers it in AWS IoT Core.
3. The Mender [Configure add-on](../../09.Add-ons/10.Configure/docs.md) sets the Thing's private key and certificate in the file (`/var/lib/mender-configure/device-config.json`) on the device. We recommend creating a [Mender configuration script](../../09.Add-ons/10.Configure/01.Device-integration/docs.md) to reconfigure the AWS IoT application and restart it with these credentials.
4. From now on, device applications can read the private key and certificate from the file and communicate with AWS IoT Core.
5. If the device gets `rejected` or `dismissed`, Mender will automatically inactivate the certificate associated with the Thing in AWS IoT Core.

> 6. If the device gets `decommissioned`, the Thing and the related resources, including the Certificate, are automatically `Deleted` from AWS IoT Core.

Reviewer 2: ```suggestion
6. If the device gets `decommissioned`, the Thing and the related resources, including the certificate, are automatically `Deleted` from AWS IoT Core.
```

> ![Device lifecycle](device_lifecycle.png)

Reviewer: This schematic refers to says "IoT Hub device details"; replace "IoT Hub" with "IoT Core", please.

Reviewer: The annotations in the schematic contain a lot of text, would it be possible to make it a little more concise, for example:
* First annotation:

> The IoT Core certificate and key for the device is saved in the Mender configuration for the device.

* Third annotation:

> The IoT Core certificate and key is made available with the configure module on the device.

## Device Shadow management

> As an additional integration point, the Mender server obtains the Device Shadow of the devices from AWS IoT Core.

Reviewer: Technically, the Mender server does not "obtain" the Device Shadow, it is merely proxying API calls.
```suggestion
The Mender server integrates with IoT Core Device Shadow service.
```

This means you can see and manage the Device Shadow directly from Mender, together with everything else about the device.

![Device Shadow flow](aws-iot-device-shadow-flow.png)

The Device Shadow is available through the Mender APIs and UI.

![Device Shadow edit](device_shadow.png)

!!! Mender strips away Device Shadow metadata from the AWS IoT Device Shadow because it's not indented to be changed by users. Thus, it's not visible in the Mender UI nor API responses. They still exist in AWS IoT Core, however.

## Role Based Access Control

!!!!! Role Based Access Control is only available in the Mender Enterprise plan.
!!!!! See [the Mender features page](https://mender.io/product/features?target=_blank)
!!!!! for an overview of all Mender plans and features.

*Admin* permission is required to set up the integration,
i.e. set the AWS access keys, region and device policy name for AWS IoT Core in the Mender's Settings.

Role Based Access Control is also enforced for the Device Shadow:
* To *see* the Device Shadow you need *Read* permission to the device

> * To *edit* the Device Shadow you need *Admin* permission

Reviewer: ```suggestion
* To *edit* the Device Shadow desired state, you need *Admin* permission
```
